apm-sniffer/apm-sdk-plugin/nats-client-plugin/src/main/java/org/apache/skywalking/apm/plugin/nats/client/NatsCommons.java

```diff
@@ -0,0 +1,124 @@
+        return StringUtil.isBlank(natsMsg.getSubject()) || natsMsg.isStatusMessage();
+    }
+
+    static Optional<AbstractSpan> createEntrySpan(Message message, Connection connection) {
+        if (message.getHeaders() == null) {
+            return Optional.empty();
+        }
+
+        ContextCarrier contextCarrier = new ContextCarrier();
+        CarrierItem next = contextCarrier.items();
+        while (next.hasNext()) {
+            next = next.next();
+            if (StringUtil.isNotEmpty(next.getHeadKey())) {
+                next.setHeadValue(message.getHeaders().getFirst(next.getHeadKey()));
+            }
+        }
+        AbstractSpan span = ContextManager.createEntrySpan("Nats/Sub/" + message.getSubject(), contextCarrier);
+        addCommonTag(span, message, connection.getConnectedUrl());
+        return Optional.of(span);
+    }
+
+    static void injectCarrier(Message message) {
+        ContextCarrier contextCarrier = new ContextCarrier();
+        ContextManager.inject(contextCarrier);
+        CarrierItem next = contextCarrier.items();
+        while (next.hasNext()) {
+            next = next.next();
+            if (StringUtil.isNotEmpty(next.getHeadKey())
+                    && StringUtil.isNotEmpty(next.getHeadValue())) {
+                message.getHeaders().add(next.getHeadKey(), next.getHeadValue());
+            }
+        }
+    }
+
+    static void addCommonTag(AbstractSpan span, Message message, String connectionUrl) {
+        Optional.ofNullable(message.getReplyTo()).ifPresent(v -> span.tag(new StringTag(REPLY_TO), v));
+        Optional.ofNullable(message.getSID()).ifPresent(v -> span.tag(new StringTag(SID), v));
+        span.setComponent(ComponentsDefine.NATS);
+        SpanLayer.asMQ(span);
+        Tags.URL.set(span, connectionUrl);
+        if (message.getStatus() != null) {
+            int code = message.getStatus().getCode();
+            String statusMsg = message.getStatus().getMessage();
+            span.tag(new IntegerTag(MSG_STATE), String.valueOf(code));
+            if (StringUtil.isNotBlank(statusMsg)) {
+                span.tag(new StringTag(MSG), statusMsg);
+            }
+            if (code != 0) {
+                span.errorOccurred();
+            }
+        }
+    }
+
+    static MessageHandler buildTraceMsgHandler(MessageHandler msgHandler, Connection connection) {
+        if (msgHandler == null) {
+            return null;
+        }
+        return msg -> {
+            if (skipTrace(msg) || msg.getHeaders() == null) {
+                msgHandler.onMessage(msg);
+                return;
+            }
+            try {
+                NatsCommons.createEntrySpan(msg, connection);
+                msgHandler.onMessage(msg);
+                ContextManager.stopSpan();
+            } catch (Exception e) {
+                if (ContextManager.isActive()) {
+                    AbstractSpan span = ContextManager.activeSpan().log(e).errorOccurred();
```

Comment:
Which span would be active in this case?

Reply:
If createEntrySpan throw a exception , I need detect whether span is active .
If    onMessage throw a exception , I need't detect it

Reply:
> Which span would be active in this case?

createEntrySpan

Reply:
How could `createEntrySpan` make an exception? What is the exception actually?

Reply:
Only my  conservative thinking .

Reply:
If we can't know there is an expected exception in a specific case, we should say there is no `try/catch`. The agent kernel would catch and log out, injecting the error into another span's logs would be very confused.

Reply:
OK, I will remove `ContextManager.isActive()` in catch . 
And I found a  serious mistake   , I should throw the exception after record this error information

Reply:
Don't worry, you just don't need `try/catch` after all.

Reply:
```
            try {
                NatsCommons.createEntrySpan(msg, connection);
                msgHandler.onMessage(msg);
                ContextManager.stopSpan();
            } catch (Exception e) {
                AbstractSpan span = ContextManager.activeSpan().log(e).errorOccurred();
                ContextManager.stopSpan(span);
                throw e;
            }
```

Reply:
> you just don't need `try/catch` after all.

I don't understand your means , If user code throw exception , it's better  that  plugin record this error message

Reply:
For this case, the codes should be like
```
            span = NatsCommons.createEntrySpan(msg, connection);
            try {
                msgHandler.onMessage(msg);
            } catch (Exception e) {
                span.log(e).errorOccurred();
                throw e;
            } finally {
                ContextManager.stopSpan(span);
            }
```

Reply:
Got it , your point at `after all ` ,